proposed/2022/dotnet-nuget-config-spec.md

```diff
@@ -0,0 +1,366 @@
+
+packageSourceMapping:
+  clear                                                            file: C:\Users\username\AppData\Roaming\NuGet\NuGet.Config
+  packageSource key="source1"  pattern="microsoft.*","nuget.*"     file: C:\Test\Repos\Solution\NuGet.Config
+  packageSource key="source2"  pattern="system.*"                  file: C:\Test\Repos\NuGet.Config                                 
+
+packageRestore:
+  add key="enabled" value="False"                                 file: C:\Users\username\AppData\Roaming\NuGet\NuGet.Config
+  add key="automatic" value="False"                               file: C:\Users\username\AppData\Roaming\NuGet\NuGet.Config
+
+```
+
+- Get `http_proxy` from config section, when invoking NuGet command in the current directory.
+
+```
+dotnet nuget config get http_proxy 
+
+http://company-squid:3128@contoso.com"
+
+```
+
+- Get `http_proxy` from config section, when invoking NuGet command in the current directory. Show the source(nuget configuration file path) of this configuration setting.
+
+```
+dotnet nuget config get http_proxy 
+
+http://company-squid:3128@contoso.com"         file: C:\Test\Repos\Solution\NuGet.Config
+
+```
+
+- Get `http_proxy` from config section, when `http_proxy` is not set in any of the NuGet configuration files.
+
+```
+dotnet nuget config get http_proxy 
+
+Key 'http_proxy' not found.
+
+```
+
+### Set
+
+Set the NuGet configuration settings. 
+
+This command will set the value of a specified NuGet configuration setting.
+
+Please note this command only manages settings in [config section](https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file#config-section).
+For other settings not in config section, we have/will have other dedicated commands. E.g., for [trustedSigners section](https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file#trustedsigners-section), we have [dotnet nuget trust](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-nuget-trust) command.
+
+#### Arguments
+
+- CONFIG_KEY
+
+Specify the key of the settings that are to be set.
+
+If the specified `CONFIG_KEY` is not a key in [config section](https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file#config-section), a message is displayed indicating the `CONFIG_KEY` could not be set.
+
+- CONFIG_VALUE
+
+Set the value of the CONFIG_KEY to CONFIG_VALUE.
+
+#### Options
+
+- --configfile <FILE>
+
+The NuGet configuration file (nuget.config) to use. If specified, only the settings from this file will be used. If it's not specified, `%AppData%\NuGet\NuGet.Config` (Windows), or `~/.nuget/NuGet/NuGet.Config` or `~/.config/NuGet/NuGet.Config` (Mac/Linux) is used. See [On Mac/Linux, the user-level config file location varies by tooling.](https://learn.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior#on-maclinux-the-user-level-config-file-location-varies-by-tooling)
```

Comment:
So, even if there's a nuget.config in the current directory, if `--configfile` is not specified, the local directory file will not be modified?

While I think it's good that there's an easy way to modify the user-profile nuget.config, I feel like there will be more customers who expect the local directory file to be edited by default, so I think this will be non-intuitive.

Reply:
Yes. This is the behavior in NuGet.exe config command right now. 
You may refer to the [doc ](https://learn.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-config#options), or the [code](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs#L208-L230) .
The code indicates if there is no clear tag, default to furthest from the user, which is the user-wide configuration file (machine wide configuration files are excluded as they have `IsReadOnly` flag set to true, so they are not writable).
Unless this is not the expected behavior in NuGet.exe, otherwise we've discussed to keep the behavior consistent between NuGet.exe and dotnet.exe, and we only consider change the behavior when major version changes, if there are enough upvotes. (Please refer to [Future Work  #2](https://github.com/NuGet/Home/blob/779cfc85bdff9887f47ae36d5705151667a80dcb/proposed/2022/dotnet-nuget-config-spec.md#future-work) in the spec).